### PR TITLE
Add HTML static code analysis report generation

### PR DESCRIPTION
**Pull request by Xenowa**
## Purpose

- Add implementations for generating an HTML static code analysis report from the scan tool

scan-report

### CODE REVIEW

File: ScanCommand/src/main/java/io/ballerina/scan/internal/ScanCmd.java [hunk 1]
```diff
@@ -150,6 +150,15 @@ public void execute() {
             outputStream.println();
             outputStream.println("View scan results at:");
             outputStream.println("\t" + reportPath + System.lineSeparator());
+
+            if (scanReport) {
+                Path scanReportPath;
+                outputStream.println("Generating scan report...");
+                scanReportPath = ScanUtils.generateScanReport(issues, project.get(), targetDir);
```
Comment (poorna2152) [inline ScanCommand/src/main/java/io/ballerina/scan/internal/ScanCmd.java:157]:
```suggestion
                outputStream.println("Generating scan report...");
                Path scanReportPath = ScanUtils.generateScanReport(issues, project.get(), targetDir);
```

File: ScanCommand/src/main/java/io/ballerina/scan/utils/ScanUtils.java [hunk 2]
```diff
@@ -120,6 +130,124 @@ private static Target getTargetPath(Project project, String directoryName) {
...
+                    throw new RuntimeException(ex);
+                }
+                jsonScanReportFile.addProperty("fileContent", fileContent);
+
+                JsonObject jsonScanReportIssueTextRange = new JsonObject();
+                LineRange lineRange = issueIml.location().lineRange();
+                jsonScanReportIssueTextRange.addProperty("startLine", lineRange.startLine().line());
+                jsonScanReportIssueTextRange.addProperty("startLineOffset", lineRange.startLine().offset());
+                jsonScanReportIssueTextRange.addProperty("endLine", lineRange.endLine().line());
+                jsonScanReportIssueTextRange.addProperty("endLineOffset", lineRange.endLine().offset());
```
Comment (poorna2152) [inline ScanCommand/src/main/java/io/ballerina/scan/utils/ScanUtils.java:167]: Shall we extract this out to a function because this is done in both `if` and `else` clauses.

File: ScanCommand/src/main/java/io/ballerina/scan/utils/ScanUtils.java [hunk 3]
```diff
@@ -120,6 +130,124 @@ private static Target getTargetPath(Project project, String directoryName) {
...
+        jsonScanReportPathAndFile.values().forEach(jsonScannedFiles::add);
+        jsonScannedProject.add("scannedFiles", jsonScannedFiles);
+        Gson gson = new GsonBuilder().setPrettyPrinting().create();
+        String jsonOutput = gson.toJson(jsonScannedProject);
+        Target target = getTargetPath(project, directoryName);
+        InputStream innerJarStream = ScanUtils.class.getResourceAsStream("/report.zip");
+
+        File htmlFile;
+        try {
+            unzipReportResources(innerJarStream, target.getReportPath().toFile());
```
Comment (poorna2152) [inline ScanCommand/src/main/java/io/ballerina/scan/utils/ScanUtils.java:213]: Let's extract the `target.getReportPath()` in to a variable

File: ScanCommand/src/main/java/io/ballerina/scan/utils/ScanUtils.java [hunk 4]
```diff
@@ -120,6 +130,124 @@ private static Target getTargetPath(Project project, String directoryName) {
...
+                try (OutputStream targetStream = new FileOutputStream(nextFile)) {
+                    final int bufferSize = 4 * 1024;
+                    final byte[] buffer = new byte[bufferSize];
+                    int nextCount;
+                    while ((nextCount = zipStream.read(buffer)) >= 0) {
+                        targetStream.write(buffer, 0, nextCount);
+                    }
+                }
+            }
+        }
```
Comment (poorna2152) [inline ScanCommand/src/main/java/io/ballerina/scan/utils/ScanUtils.java:248]:
```suggestion
        while ((nextEntry = zipStream.getNextEntry()) != null) {
          if (nextEntry.isDirectory()) {
              continue;
          }
          final File nextFile = new File(target, nextEntry.getName());
          final File parent = nextFile.getParentFile();
      
          if (parent != null) {
              Files.createDirectories(parent.toPath());
          }
          try (OutputStream targetStream = new FileOutputStream(nextFile)) {
              final int bufferSize = 4 * 1024;
              final byte[] buffer = new byte[bufferSize];
              int nextCount = zipStream.read(buffer);
              while (nextCount >= 0) {
                  targetStream.write(buffer, 0, nextCount);
                  nextCount = zipStream.read(buffer);
              }
          }
      }

```

File: ScanCommand/src/main/java/io/ballerina/scan/internal/ScanToolConstants.java [hunk 5]
```diff
@@ -27,8 +27,11 @@ public class ScanToolConstants {
     static final String SCAN_COMMAND = "scan";
     public static final String RESULTS_JSON_FILE = "scan_results.json";
     static final String BALLERINA_RULE_PREFIX = "B";
-    public static final String PATH_SEPARATOR = "/";
-    public static final String BALLERINA_ORG = "ballerina";
+    static final String PATH_SEPARATOR = "/";
```
Comment (SasinduDilshara) [inline ScanCommand/src/main/java/io/ballerina/scan/internal/ScanToolConstants.java:30]: Shall we use a inbuilt function to get this

File: ScanCommand/src/main/java/io/ballerina/scan/internal/ScanToolConstants.java [hunk 6]
```diff
@@ -27,8 +27,11 @@ public class ScanToolConstants {
...
+    static final String BALLERINA_ORG = "ballerina";
```
Comment (SasinduDilshara) [inline ScanCommand/src/main/java/io/ballerina/scan/internal/ScanToolConstants.java:31]: Is this a `ballerina` module?. isnt it `ballerinai` or `ballerinax`.

File: ScanCommand/src/main/java/io/ballerina/scan/utils/ScanUtils.java [hunk 7]
```diff
@@ -120,6 +130,124 @@ private static Target getTargetPath(Project project, String directoryName) {
...
+     * Returns the {@link Path} of the html analysis report where generated issues are saved.
+     *
+     * @param issues        generated issues
+     * @param project       Ballerina project
+     * @param directoryName target directory name
+     * @return path of the html analysis report where generated issues are saved
+     */
+    public static Path generateScanReport(List<Issue> issues, Project project, String directoryName) {
+        JsonObject jsonScannedProject = new JsonObject();
+        jsonScannedProject.addProperty("projectName", project.currentPackage().packageName().toString());
```
Comment (SasinduDilshara) [inline ScanCommand/src/main/java/io/ballerina/scan/utils/ScanUtils.java:143]: These keys should be constants.

File: ScanCommand/src/test/java/io/ballerina/scan/internal/ScanCmdTest.java [hunk 8]
```diff
@@ -153,4 +179,67 @@ void testScanCommandSingleFileProjectWithTooManyArguments() throws IOException {
...
+        issues.add(new IssueImpl(location, coreRule, Source.BUILT_IN, "main.bal",
+                validBalProject.resolve("main.bal").toString()));
+        issues.add(new IssueImpl(location, externalRule, Source.EXTERNAL, "main.bal",
+                validBalProject.resolve("main.bal").toString()));
+        String userDir = System.getProperty("user.dir");
+        System.setProperty("user.dir", validBalProject.toString());
+        Project project = ProjectLoader.loadProject(validBalProject);
+        System.setProperty("user.dir", userDir);
+        Path resultsFile = ScanUtils.saveToDirectory(issues, project, null);
+        String result = Files.readString(resultsFile, StandardCharsets.UTF_8).replace("\r\n", "\n");
```
Comment (SasinduDilshara) [inline ScanCommand/src/test/java/io/ballerina/scan/internal/ScanCmdTest.java:200]: make this line separates as constant variables in the class

File: ScanCommand/src/main/java/io/ballerina/scan/internal/ScanToolConstants.java (hunk 6 above)
Comment (gayaldassanayake) [reply ScanCommand/src/main/java/io/ballerina/scan/internal/ScanToolConstants.java:31]: bal tools should be of the orgs `ballerina`, `ballerinax`. So far we have, 
1. edi tool - `ballerina`
2. copybook - `ballerina`
3. health - `ballerinax`

We can also use one of those. Since this does not really an external package, `ballerina` is fine IMO.

looping @MaryamZi for more insight.

File: ScanCommand/src/main/java/io/ballerina/scan/internal/ScanCmd.java [hunk 9]
```diff
@@ -150,6 +150,15 @@ public void execute() {
...
+                outputStream.println();
+                outputStream.println("View scan report at:");
+                outputStream.println("\t" + ScanToolConstants.FILE_PROTOCOL + scanReportPath + System.lineSeparator());
```
Comment (gayaldassanayake) [inline ScanCommand/src/main/java/io/ballerina/scan/internal/ScanCmd.java:160]: Can use multiline strings like,

```suggestion
                outputStream.println("""
                      View scan report at:
                          ${ScanToolConstants.FILE_PROTOCOL + scanReportPath}
                      """);
```

File: ScanCommand/src/main/java/io/ballerina/scan/utils/ScanUtils.java [hunk 10]
```diff
@@ -120,6 +130,124 @@ private static Target getTargetPath(Project project, String directoryName) {
...
+            }
+            jsonScanReportPathAndFile.put(filePath, jsonScanReportFile);
+        });
+        JsonArray jsonScannedFiles = new JsonArray();
+        jsonScanReportPathAndFile.values().forEach(jsonScannedFiles::add);
+        jsonScannedProject.add("scannedFiles", jsonScannedFiles);
+        Gson gson = new GsonBuilder().setPrettyPrinting().create();
+        String jsonOutput = gson.toJson(jsonScannedProject);
+        Target target = getTargetPath(project, directoryName);
+        InputStream innerJarStream = ScanUtils.class.getResourceAsStream("/report.zip");
```
Comment (gayaldassanayake) [inline ScanCommand/src/main/java/io/ballerina/scan/utils/ScanUtils.java:209]: Check if `/report.zip` is working fine in both unix and windows.

File: ScanCommand/src/main/java/io/ballerina/scan/internal/ScanToolConstants.java (hunk 5 above)
Comment (Xenowa) [reply ScanCommand/src/main/java/io/ballerina/scan/internal/ScanToolConstants.java:30]: This constant was meant to be used as a separator for creating the fully qualified rule Identifiers for rules introduced via compiler plugins (i.e.: ballerina/io:B107). Will reword it to something else.

File: ScanCommand/src/main/java/io/ballerina/scan/internal/ScanToolConstants.java (hunk 6 above)
Comment (Xenowa) [reply ScanCommand/src/main/java/io/ballerina/scan/internal/ScanToolConstants.java:31]: This constant was meant to be used to label the fully qualified rule identifiers for non-core rules introduced via compiler plugins, specifically compiler plugins developed by  Ballerina platform developers, so that the resulting core rules and external rules are represented as:
1. Core rule: `B101`
2. ballerina/io compiler plugin rule: `ballerina/io:B101`

As `ballerinai` as well as `ballerinax` compiler plugins are also created by Ballerina platform developers, they will be added as constants as well.

File: ScanCommand/src/main/java/io/ballerina/scan/utils/ScanUtils.java [hunk 11]
```diff
@@ -120,6 +130,124 @@ private static Target getTargetPath(Project project, String directoryName) {
...
+            JsonObject jsonScanReportFile;
+            if (!jsonScanReportPathAndFile.containsKey(filePath)) {
+                jsonScanReportFile = new JsonObject();
+                jsonScanReportFile.addProperty("fileName", issueImpl.fileName());
+                jsonScanReportFile.addProperty("filePath", filePath);
+                String fileContent = "";
+                try {
+                    fileContent = Files.readString(Path.of(filePath));
+                } catch (IOException ex) {
+                    throw new RuntimeException(ex);
```
Comment (gayaldassanayake) [inline ScanCommand/src/main/java/io/ballerina/scan/utils/ScanUtils.java:158]: Let's append a message to the exception message to make it more context-sensitive. 
```suggestion
                    throw new RuntimeException("Failed to read the file with exception: " + ex.getMessage());
```

Comment (gayaldassanayake) [reply ScanCommand/src/main/java/io/ballerina/scan/utils/ScanUtils.java:158]: Follow the same approach for other places as well.

File: ScanCommand/src/main/java/io/ballerina/scan/internal/ScanCmd.java (hunk 9 above)
Comment (Xenowa) [reply ScanCommand/src/main/java/io/ballerina/scan/internal/ScanCmd.java:160]: This does not seem to work with Java 17, will be using the previous approach.

File: ScanCommand/src/main/java/io/ballerina/scan/internal/ScanToolConstants.java (hunk 6 above)
Comment (Xenowa) [reply ScanCommand/src/main/java/io/ballerina/scan/internal/ScanToolConstants.java:30]: After discussions with @shafreenAnfar, since the scan tool is not tied with an external product, it was decided that it should go under `ballerina` org.

File: StaticCodeAnalysisReport/scripts/update-version.js [hunk 12]
```diff
@@ -0,0 +1,8 @@
+const fs = require("fs");
+const path = require("path");
+const packageJson = require("../package.json");
+packageJson.version = process.argv[2];
+
+fs.writeFileSync(
+    path.join(__dirname, "..", "package.json"),
+    JSON.stringify(packageJson, null, 4)+"\n");
```
Comment (gayaldassanayake) [inline StaticCodeAnalysisReport/scripts/update-version.js:8]: If this is not an auto generated file, let's add a new line

File: StaticCodeAnalysisReport/src/App.jsx [hunk 13]
```diff
@@ -0,0 +1,54 @@
+import Header from "./components/Header";
+import { useState } from "react";
+import MainView from "./components/MainView";
+import SingleFileView from "./components/SingleFileView";
+
```
Comment (gayaldassanayake) [inline StaticCodeAnalysisReport/src/App.jsx:5]: Let's keep only one blank line

File: StaticCodeAnalysisReport/src/App.jsx [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+      <Header projectName={analysisResults.projectName} />
+      {fileView.viewToggle ?
+        <SingleFileView toggleMainView={toggleMainView} requestedFile={fileView.requestedFile} /> :
+        <MainView toggleSingleFileView={toggleSingleFileView} analyzedFiles={analysisResults.scannedFiles} />
+      }
+    </>
+  );
+}
+
+export default App
```
Comment (gayaldassanayake) [inline StaticCodeAnalysisReport/src/App.jsx:54]: New line. Check for other places as well

**Review by gayaldassanayake — APPROVED** [review]

File: ScanCommand/src/main/java/io/ballerina/scan/internal/ReporterImpl.java [hunk 15]
```diff
@@ -72,18 +74,22 @@ private Issue createIssue(Document reportedDocument, Location location, Rule rul
...
         String[] parts = fullyQualifiedRuleId.split(":");
+
         Source source;
         if (parts[0].equals(BALLERINA_RULE_PREFIX + rule.numericId())) {
             source = Source.BUILT_IN;
         } else {
             String reportedSource = parts[0];
-            String pluginOrg = reportedSource.split(PATH_SEPARATOR)[0];
-            source = pluginOrg.equals(BALLERINA_ORG) ? Source.BUILT_IN : Source.EXTERNAL;
+            String pluginOrg = reportedSource.split(PACKAGE_ORG_NAME_SEPARATOR)[0];
```
Comment (MaryamZi) [inline ScanCommand/src/main/java/io/ballerina/scan/internal/ReporterImpl.java:85]: Is `reportedSource` the correct name here? Reads more like the source for which the issue was reported.

File: ScanCommand/src/main/java/io/ballerina/scan/internal/ScanCmd.java [hunk 16]
```diff
@@ -145,11 +145,18 @@ public void execute() {
...
         if (project.get().kind().equals(ProjectKind.BUILD_PROJECT)) {
-            Path reportPath;
-            reportPath = ScanUtils.saveToDirectory(issues, project.get(), targetDir);
+            Path reportPath = ScanUtils.saveToDirectory(issues, project.get(), targetDir);
             outputStream.println();
             outputStream.println("View scan results at:");
             outputStream.println("\t" + reportPath + System.lineSeparator());
+
+            if (scanReport) {
+                outputStream.println("Generating scan report...");
```
Comment (MaryamZi) [inline ScanCommand/src/main/java/io/ballerina/scan/internal/ScanCmd.java:154]: Do we really need these kinda logs? Not necessary and don't add value IMO.

File: ScanCommand/src/main/java/io/ballerina/scan/utils/ScanUtilConstants.java [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+ *  software distributed under the License is distributed on an
+ *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ *  KIND, either express or implied. See the License for the
+ *  specific language governing permissions and limitations
+ *  under the License.
+ */
+
+package io.ballerina.scan.utils;
+
+public class ScanUtilConstants {
```
Comment (MaryamZi) [inline ScanCommand/src/main/java/io/ballerina/scan/utils/ScanUtilConstants.java:21]: This is already under `io.ballerina.scan.utils`, so including `ScanUtil` feels redundant.

File: ScanCommand/src/main/java/io/ballerina/scan/utils/ScanUtils.java [hunk 18]
```diff
@@ -120,6 +147,129 @@ private static Target getTargetPath(Project project, String directoryName) {
...
+     * @param issues        generated issues
+     * @param project       Ballerina project
+     * @param directoryName target directory name
+     * @return path of the html analysis report where generated issues are saved
+     */
+    public static Path generateScanReport(List<Issue> issues, Project project, String directoryName) {
+        JsonObject jsonScannedProject = new JsonObject();
+        jsonScannedProject.addProperty(SCAN_REPORT_PROJECT_NAME, project.currentPackage().packageName().toString());
+        Map<String, JsonObject> jsonScanReportPathAndFile = new HashMap<>();
+        issues.forEach((issue) -> {
```
Comment (MaryamZi) [inline ScanCommand/src/main/java/io/ballerina/scan/utils/ScanUtils.java:162]: I'd use a simple for loop and avoid streams.

File: ScanCommand/src/main/java/io/ballerina/scan/utils/ScanUtils.java [hunk 19]
```diff
@@ -120,6 +147,129 @@ private static Target getTargetPath(Project project, String directoryName) {
...
+            IssueImpl issueImpl = (IssueImpl) issue;
+            String filePath = issueImpl.filePath();
+
+            JsonObject jsonScanReportFile;
+            if (!jsonScanReportPathAndFile.containsKey(filePath)) {
```
Comment (MaryamZi) [inline ScanCommand/src/main/java/io/ballerina/scan/utils/ScanUtils.java:167]: Can also then check the inverse and continue and minimize indentation.

File: ScanCommand/src/main/java/io/ballerina/scan/utils/ScanUtils.java [hunk 20]
```diff
@@ -120,6 +147,129 @@ private static Target getTargetPath(Project project, String directoryName) {
...
+     *
+     * @param issues        generated issues
+     * @param project       Ballerina project
+     * @param directoryName target directory name
+     * @return path of the html analysis report where generated issues are saved
+     */
+    public static Path generateScanReport(List<Issue> issues, Project project, String directoryName) {
+        JsonObject jsonScannedProject = new JsonObject();
+        jsonScannedProject.addProperty(SCAN_REPORT_PROJECT_NAME, project.currentPackage().packageName().toString());
+        Map<String, JsonObject> jsonScanReportPathAndFile = new HashMap<>();
```
Comment (MaryamZi) [inline ScanCommand/src/main/java/io/ballerina/scan/utils/ScanUtils.java:161]: Including json in variable names may be redundant. Will help shorten the names too.

File: ScanCommand/src/test/java/io/ballerina/scan/TestConstants.java [hunk 21]
```diff
@@ -0,0 +1,26 @@
...
+ *  specific language governing permissions and limitations
+ *  under the License.
+ */
+
+package io.ballerina.scan;
+
+public class TestConstants {
+    public static final String WINDOWS_LINE_SEPARATOR = "\r\n";
+    public static final String LINUX_LINE_SEPARATOR = "\n";
+    private TestConstants() {
```
Comment (MaryamZi) [inline ScanCommand/src/test/java/io/ballerina/scan/TestConstants.java:24]:
```suggestion

    private TestConstants() {
```

File: settings.gradle [hunk 22]
```diff
@@ -30,3 +30,5 @@ pluginManagement {
 
 rootProject.name = 'static-code-analysis-tool'
 include 'ScanCommand'
+include 'StaticCodeAnalysisReport'
+
```
Comment (MaryamZi) [inline settings.gradle:34]:
```suggestion
```
